### PR TITLE
[evm] EVM upgrade to cancun

### PR DESCRIPTION
**Pull request by millken**
# Description
evm upgrade to v1.13.12
change big.Int to uint256.Int
golang version upgrade to 1.21

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -25,7 +25,7 @@ jobs:
       - name: Set up Go
         uses: actions/setup-go@v3
         with:
-          go-version: 1.19.12
+          go-version: 1.21.4
```
Comment (millken) [inline .github/workflows/ci.yaml:28]: because go-ethereum has been upgraded to 1.21

File: Makefile [hunk 2]
```diff
@@ -11,7 +11,7 @@ GOVET=$(GOCMD) vet
 GOBUILD=$(GOCMD) build
 GOINSTALL=$(GOCMD) install
 GOCLEAN=$(GOCMD) clean
-GOTEST=$(GOCMD) test
+GOTEST=GOARCH=amd64 CGO_ENABLED=1 $(GOCMD) test
```
Comment (millken) [inline Makefile:14]: fix test failed in mac m1

File: action/protocol/execution/evm/evm_test.go [hunk 3]
```diff
@@ -356,8 +356,8 @@ func TestConstantinople(t *testing.T) {
 		require.Equal(isSumatra, chainRules.IsShanghai)
 
 		// Cancun, Prague not yet enabled
-		require.False(evmChainConfig.IsCancun(evm.Context.Time))
-		require.False(evmChainConfig.IsPrague(evm.Context.Time))
+		require.False(evmChainConfig.IsCancun(big.NewInt(int64(e.height)), evm.Context.Time))
```
Comment (millken) [inline action/protocol/execution/evm/evm_test.go:359]: the first parameter is block height

File: blockindex/contractstaking/indexer_test.go [hunk 4]
```diff
@@ -884,8 +885,8 @@ func TestContractStakingIndexerVotes(t *testing.T) {
 		bts, err := indexer.BucketTypes(height)
 		r.NoError(err)
 		r.Len(bts, 4)
-		slices.SortFunc(bts, func(i, j *staking.ContractStakingBucketType) bool {
-			return i.Amount.Int64() < j.Amount.Int64()
+		slices.SortFunc(bts, func(i, j *staking.ContractStakingBucketType) int {
```
Comment (millken) [inline blockindex/contractstaking/indexer_test.go:888]: SortFunc has been changed in go1.21

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (dustinxie) [reply .github/workflows/ci.yaml:28]: no need, let's keep 1.19

File: action/protocol/execution/evm/evm.go [hunk 5]
```diff
@@ -89,7 +90,7 @@ type (
 		context     vm.BlockContext
 		txCtx       vm.TxContext
 		nonce       uint64
-		amount      *big.Int
+		amount      *uint256.Int
```
Comment (dustinxie) [inline action/protocol/execution/evm/evm.go:93]: no need to change

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 6]
```diff
@@ -49,7 +50,7 @@ type (
 		blockHeight                uint64
 		executionHash              hash.Hash256
 		lastAddBalanceAddr         string
-		lastAddBalanceAmount       *big.Int
+		lastAddBalanceAmount       *uint256.Int
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:53]: no need to change

File: action/protocol/execution/evm/evmstatedbadapter_test.go [hunk 7]
```diff
@@ -319,7 +319,7 @@ func TestNonce(t *testing.T) {
 var tests = []stateDBTest{
 	{
 		[]bal{
-			{_addr1, big.NewInt(40000)},
+			{_addr1, uint256.NewInt(40000)},
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter_test.go:322]: no need to change, just convert to uint256 when it is used

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (millken) [reply .github/workflows/ci.yaml:28]:
```
Error: ../../../go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20230928194634-aa077af62593/vfs/mem_fs.go:469:29: y.lockedFiles.Swap undefined (type sync.Map has no field or method Swap)
221note: module requires Go 1.20
```

File: action/protocol/execution/evm/evm.go (hunk 5 above)
Comment (millken) [reply action/protocol/execution/evm/evm.go]: reverted

File: action/protocol/execution/evm/evmstatedbadapter.go (hunk 6 above)
Comment (millken) [reply action/protocol/execution/evm/evmstatedbadapter.go]: reverted

File: action/protocol/execution/evm/evmstatedbadapter_test.go (hunk 7 above)
Comment (millken) [reply action/protocol/execution/evm/evmstatedbadapter_test.go]: reverted

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (envestcc) [reply .github/workflows/ci.yaml:28]: also required in Dockerfile

File: Makefile (hunk 2 above)
Comment (envestcc) [reply Makefile:14]: Is this flag `-gcflags=all=-l` not requried? It's noted in gomonkey doc: https://github.com/agiledragon/gomonkey#notes

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 8]
```diff
@@ -485,6 +487,11 @@ func (stateDB *StateDBAdapter) GetTransientState(addr common.Address, key common
 	return common.Hash{}
 }
 
+// Selfdestruct6780 implements EIP-6780
+func (stateDB *StateDBAdapter) Selfdestruct6780(evmAddr common.Address) {
+	//Todo: implement EIP-6780
```
Comment (envestcc) [inline action/protocol/execution/evm/evmstatedbadapter.go:492]: panic() now?

File: action/protocol/poll/nativestaking.go [hunk 9]
```diff
@@ -145,7 +145,7 @@ func (ns *NativeStaking) readBuckets(ctx context.Context, prevIndx, limit *big.I
 	// decode the contract read result
 	res, err := ns.abi.Unpack("getActivePyggs", data)
 	if err != nil {
-		if err.Error() == "abi: attempting to unmarshall an empty string while arguments are expected" {
+		if err.Error() == "abi: attempting to unmarshal an empty string while arguments are expected" {
```
Comment (envestcc) [inline action/protocol/poll/nativestaking.go:148]: does it need a hard-fork?

File: go.mod [hunk 10]
```diff
@@ -226,14 +240,13 @@ require (
...
 	golang.org/x/term v0.15.0 // indirect
-	golang.org/x/time v0.0.0-20220922220347-f3bd1da661af // indirect
-	gopkg.in/natefinch/npipe.v2 v2.0.0-20160621034901-c1b8fa8bdcce // indirect
+	golang.org/x/time v0.3.0 // indirect
 	gopkg.in/square/go-jose.v2 v2.5.1 // indirect
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
 
-replace github.com/ethereum/go-ethereum => github.com/iotexproject/go-ethereum v1.7.4-0.20230806203205-6819e8158a5f
+replace github.com/ethereum/go-ethereum => github.com/iotexproject/go-ethereum v1.7.4-0.20240321181400-fc0405c7567c
```
Comment (envestcc) [inline go.mod:250]: replace with tag before merge

File: Makefile (hunk 2 above)
Comment (millken) [reply Makefile:14]: > Is this flag `-gcflags=all=-l` not requried? It's noted in gomonkey doc: https://github.com/agiledragon/gomonkey#notes

Makefile has been added before 
```
-gcflags="all=-N -l"
```

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (millken) [reply .github/workflows/ci.yaml:28]: done

File: action/protocol/execution/evm/evmstatedbadapter.go (hunk 8 above)
Comment (millken) [reply action/protocol/execution/evm/evmstatedbadapter.go:492]: done

File: go.mod (hunk 10 above)
Comment (millken) [reply go.mod:250]: Wait until the EIP is completed, and then tag it at the end.

File: action/protocol/poll/nativestaking.go (hunk 9 above)
Comment (envestcc) [reply action/protocol/poll/nativestaking.go:148]: add TODO

File: go.mod (hunk 10 above)
Comment (envestcc) [reply go.mod:250]: add TODO

**Review by envestcc — APPROVED** [review]

File: action/protocol/execution/evm/evm.go [hunk 11]
```diff
@@ -493,7 +494,7 @@ func executeInEVM(evmParams *Params, stateDB *StateDBAdapter) ([]byte, uint64, u
 	if evmParams.contract == nil {
 		// create contract
 		var evmContractAddress common.Address
-		_, evmContractAddress, remainingGas, evmErr = evm.Create(executor, evmParams.data, remainingGas, evmParams.amount)
+		_, evmContractAddress, remainingGas, evmErr = evm.Create(executor, evmParams.data, remainingGas, uint256.MustFromBig(evmParams.amount))
```
Comment (dustinxie) [inline action/protocol/execution/evm/evm.go:497]: add `amount = uint256.MustFromBig(evmParams.amount)` to L493

File: action/protocol/execution/evm/evm.go [hunk 12]
```diff
@@ -503,7 +504,7 @@ func executeInEVM(evmParams *Params, stateDB *StateDBAdapter) ([]byte, uint64, u
 	} else {
 		stateDB.SetNonce(evmParams.txCtx.Origin, stateDB.GetNonce(evmParams.txCtx.Origin)+1)
 		// process contract
-		ret, remainingGas, evmErr = evm.Call(executor, *evmParams.contract, evmParams.data, remainingGas, evmParams.amount)
+		ret, remainingGas, evmErr = evm.Call(executor, *evmParams.contract, evmParams.data, remainingGas, uint256.MustFromBig(evmParams.amount))
```
Comment (dustinxie) [inline action/protocol/execution/evm/evm.go:507]: use `amount` here

File: action/protocol/poll/nativestaking.go (hunk 9 above)
Comment (dustinxie) [reply action/protocol/poll/nativestaking.go:148]: revert this change, do it in another PR

**Review by dustinxie — APPROVED** [review]

File: action/protocol/poll/nativestaking.go (hunk 9 above)
Comment (dustinxie) [reply action/protocol/poll/nativestaking.go:148]: pls fix before merge

Comment (millken) [reply action/protocol/poll/nativestaking.go:148]: it's a typo error fixed by goethereum, if revert, ci will not pass

**Pull request merged**